File: connectors/javascript/tests/integration/advanced/A03_SendMethod.test.ts
/**
 * A-03: Send() Method Tests (Fire-and-Forget)
 *
 * Tests for the send() method which sends messages without waiting for responses.
 * Unlike request(), send() is used for one-way messages like notifications or events.
 */

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { BaseIntegrationTest } from '../helpers/BaseIntegrationTest.js';
import { Packet } from '../../../src/packet.js';

describe('A-03: Send Method (Fire-and-Forget)', () => {
    let testContext: BaseIntegrationTest;

    beforeEach(async () => {
        testContext = new BaseIntegrationTest();
        await testContext['beforeEach']();
    });

    afterEach(async () => {
        await testContext['afterEach']();
    });

    test('A-03-01: send() sends message successfully', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('send-test-user-1');

        // When: Send message using send() (fire-and-forget)
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);

        // Then: No exception should be thrown and connection maintained
        await testContext['delay'](100);
        expect(testContext['connector']!.isConnected).toBe(true);
    });

    test('A-03-02: Connection is maintained after send()', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('send-test-user-2');

        // When: Send 10 messages
        for (let i = 0; i < 10; i++) {
            const echoRequest = Packet.empty('EchoRequest');
            testContext['connector']!.send(echoRequest);
        }

        await testContext['delay'](200);

        // Then: Connection should still be maintained
        expect(testContext['connector']!.isConnected).toBe(true);
        expect(testContext['connector']!.isAuthenticated).toBe(true);
    });

    test('A-03-03: send() and request() can be mixed', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('send-test-user-3');

        // When: Mix send() and request() calls
        // Send some messages
        for (let i = 0; i < 5; i++) {
            const sendRequest = Packet.empty('EchoRequest');
            testContext['connector']!.send(sendRequest);
        }

        // Make a request in between
        const echoReply = await testContext['echo']('Request in between', 100);

        // Send more messages
        for (let i = 5; i < 10; i++) {
            const sendRequest = Packet.empty('EchoRequest');
            testContext['connector']!.send(sendRequest);
        }

        // Then: Request should complete successfully
        expect(echoReply.content).toBe('Request in between');
        expect(testContext['connector']!.isConnected).toBe(true);
    });

    test('A-03-04: send() with BroadcastRequest triggers push message', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('send-test-user-4');

        let receivedPush = false;
        let pushData = '';

        testContext['connector']!.onReceive = (packet: Packet) => {
            if (packet.msgId === 'BroadcastNotify') {
                receivedPush = true;
                const notify = testContext['parsePayload'](packet);
                pushData = notify.data || notify.content || 'Hello from Send!';
            }
        };

        // When: Send broadcast request
        const broadcastRequest = Packet.empty('BroadcastRequest');
        testContext['connector']!.send(broadcastRequest);

        // Wait for push message with mainThreadAction
        const received = await testContext['waitForConditionWithMainThread'](
            () => receivedPush,
            5000
        );

        // Then: Should receive push notification
        expect(received).toBe(true);
        expect(receivedPush).toBe(true);
    });

    test('A-03-05: send() after disconnect fires OnError', async () => {
        // Given: Connected and authenticated, then disconnected
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('send-test-user-5');

        let errorFired = false;
        let errorCode = 0;

        testContext['connector']!.onError = (stageId: bigint, code: number, message: string) => {
            errorFired = true;
            errorCode = code;
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Try to send after disconnect
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);
        testContext['connector']!.mainThreadAction();

        // Then: OnError should be fired
        expect(errorFired).toBe(true);
        expect(errorCode).toBeGreaterThan(0);
    });

    test('A-03-06: send() before authentication may trigger OnError', async () => {
        // Given: Connected but not authenticated
        await testContext['createStageAndConnect']();

        let errorFired = false;

        testContext['connector']!.onError = () => {
            errorFired = true;
        };

        // When: Try to send before authentication
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);
        testContext['connector']!.mainThreadAction();

        await testContext['delay'](100);

        // Then: Implementation-dependent - server may reject or allow
        // Test just verifies no crash occurs
        expect(testContext['connector']!.isConnected).toBe(true);
    });

    test('A-03-07: Rapid fire send() calls are all processed', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('send-test-user-7');

        // When: Send 100 rapid messages
        const messageCount = 100;
        for (let i = 0; i < messageCount; i++) {
            const echoRequest = Packet.empty('EchoRequest');
            testContext['connector']!.send(echoRequest);
        }

        // Give time for all messages to be sent
        await testContext['delay'](500);

        // Then: Connection should still be alive
        expect(testContext['connector']!.isConnected).toBe(true);

        // Verify server is still responsive with a request
        const echoReply = await testContext['echo']('Check', 999);
        expect(echoReply.content).toBe('Check');
    });
});

File: connectors/cpp/tests/integration/advanced/a03_send_method_test.cpp
#include "../base_integration_test.hpp"
#include <gtest/gtest.h>
#include <atomic>

using namespace playhouse;
using namespace playhouse::test;

/// A-03: Send Method Tests
/// Verifies fire-and-forget Send() method (no response expected)
class A03_SendMethodTest : public BaseIntegrationTest {};

TEST_F(A03_SendMethodTest, Send_SimpleMessage_NoResponse) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_user"));

    // When: Send message without expecting response
    Bytes payload = proto::EncodeEchoRequest("Fire and Forget", 1);
    connector_->Send(Packet::FromBytes("EchoRequest", std::move(payload)));

    // Process callbacks to ensure send completes
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: Should complete without error
    EXPECT_TRUE(connector_->IsConnected());
    SUCCEED() << "Send method completed without response";
}

TEST_F(A03_SendMethodTest, Send_MultipleConcurrent_AllSent) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_multi_user"));

    // When: Send multiple messages rapidly
    for (int i = 0; i < 10; i++) {
        Bytes payload = proto::EncodeEchoRequest("Message " + std::to_string(i), i);
        connector_->Send(Packet::FromBytes("EchoRequest", std::move(payload)));
    }

    // Process callbacks
    for (int i = 0; i < 20; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: All should be sent without blocking
    EXPECT_TRUE(connector_->IsConnected());
    SUCCEED() << "Multiple Send operations completed";
}

TEST_F(A03_SendMethodTest, Send_VsRequest_BothWork) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_mix_user"));

    // When: Mix Send and Request operations
    // Send without response
    Bytes notify_payload = proto::EncodeEchoRequest("Notification", 0);
    connector_->Send(Packet::FromBytes("EchoRequest", std::move(notify_payload)));

    // Request with response
    Bytes echo_payload = proto::EncodeEchoRequest("Echo test", 1);
    auto packet = Packet::FromBytes("EchoRequest", std::move(echo_payload));

    // Then: Request should still work after Send
    Packet response = Packet::Empty("Empty");
    bool completed = RequestAndWait(std::move(packet), response, 5000);
    ASSERT_TRUE(completed) << "Request should complete after Send";
    EXPECT_EQ(response.GetErrorCode(), 0);
    SUCCEED() << "Send and Request both work together";
}

TEST_F(A03_SendMethodTest, Send_Broadcast_Triggers_Push) {
    // Given: Connected and authenticated
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_broadcast_user"));

    std::atomic<bool> received{false};
    std::string received_data;

    connector_->OnReceive = [&](Packet packet) {
        if (packet.GetMsgId() == "BroadcastNotify") {
            std::string event_type;
            proto::DecodeBroadcastNotify(packet.GetPayload(), event_type, received_data);
            received = true;
        }
    };

    // When: Send broadcast via Send
    Bytes payload = proto::EncodeBroadcastRequest("Hello from Send!");
    connector_->Send(Packet::FromBytes("BroadcastRequest", std::move(payload)));

    // Then: Push message should be received
    bool completed = WaitForConditionWithMainThreadAction([&]() {
        return received.load();
    }, 5000);

    ASSERT_TRUE(completed);
    EXPECT_EQ(received_data, "Hello from Send!");
}

TEST_F(A03_SendMethodTest, Send_BeforeDisconnect_HandledGracefully) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_disconnect_user"));

    std::atomic<bool> error_triggered{false};

    connector_->OnError = [&](int code, std::string message) {
        if (code == error_code::CONNECTION_CLOSED) {
            error_triggered = true;
        }
    };

    // When: Disconnect then try to send
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    Bytes payload = proto::EncodeEchoRequest("AfterDisconnect", 1);
    connector_->Send(Packet::FromBytes("EchoRequest", std::move(payload)));

    // Process callbacks
    bool completed = WaitForConditionWithMainThreadAction([&]() {
        return error_triggered.load();
    }, 5000);

    // Then: Should trigger error event
    EXPECT_TRUE(completed) << "Error should be triggered for send after disconnect";
}

TEST_F(A03_SendMethodTest, Send_WithEmptyPayload_HandledCorrectly) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_empty_user"));

    // When: Send with empty payload
    Bytes empty_payload;
    connector_->Send(Packet::FromBytes("EchoRequest", std::move(empty_payload)));

    // Process callbacks
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: Should handle gracefully
    EXPECT_TRUE(connector_->IsConnected());
    SUCCEED() << "Empty payload send handled gracefully";
}

TEST_F(A03_SendMethodTest, Send_LargePayload_Succeeds) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("send_large_user"));

    // When: Send large payload via Send method
    std::string large_content(50 * 1024, 'X');  // 50KB
    Bytes payload = proto::EncodeEchoRequest(large_content, 99);
    connector_->Send(Packet::FromBytes("EchoRequest", std::move(payload)));

    // Process callbacks
    for (int i = 0; i < 50; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(20));
    }

    // Then: Should handle without error
    EXPECT_TRUE(connector_->IsConnected());
    SUCCEED() << "Large payload sent successfully";
}
